### DISCUSSION TITLE
New on-board controller and Optitrack integration

### DISCUSSION THREAD

**Discussion by xkhainguyen** [bitcraze/discussions#657] [Q&A] [answered]
Hi,

We have developed a new controller library in C (with my own linear algebra and utility code) and want to integrate it into current firmware (onboard). 

1. What is the best way to do it? Should we call it as a module/library or put all the code directly into a my_controller.c
2. We also need to send measurement from our Optitrack system (maybe full state). Does the firmware use additional filter to fuse it with on-board sensors? And what is the best practice to get high-accuracy high-frequency full-state measurement?
3. Should we use cflib or crazyswarm for that?

Thank you very much!

**Comment by krichardsson** [comment, marked as answer]
Hi!

The best option is probably to create an app (out of tree build) in the firmware with the controller. The easiest way to integrate your lib is probably to copy the lib files into your app, but this is maybe not ideal for future maintenance (depending on your needs/future plans). If you have your lib in a git repo, you can add it as a git submodule in your app (this is what we do with some libs in the Crayflie firmare), but I have never done this in an app and I'm not sure if it becomes complex with the make files.

You can find documentation on [apps](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/userguides/app_layer/) and [out-of-tree builds, including adding a controller](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/development/oot/) in our documentation.
There is an example [app with files in a sub directory](https://github.com/bitcraze/crazyflie-firmware/tree/master/examples/app_hello_file_tree) that might be useful for you.
Also see the [Adding an estimator or controller](https://www.bitcraze.io/2023/02/adding-an-estimator-or-controller/) blog post

**Reply by xkhainguyen** [reply]
Thank you very much for your response!
Our single optimal controller is expected to run at 50-100Hz, for full-state feedback from OptiTrack. Do you think that frequency is able to control the Crazyflie effectively?

Other questions have not been resolved yet --

1. We must also send measurements from our OptiTrack system (maybe full-state). Does the firmware use an additional filter to fuse it with onboard sensors? We are concerned about the communication latency then we plan to dead-reckon the OptiTrack measurement using IMU, what do you think?
2. And what is the best practice to get the high-accuracy high-frequency full-state measurement?
3. Should we use cflib or crazyswarm for that?
4. Is there possible to build a very optimized program (no assertion, printing, overhead, etc)?

**Comment by krichardsson** [comment]
> We must also send measurements from our OptiTrack system (maybe full-state). Does the firmware use an additional filter to fuse it with onboard sensors? We are concerned about the communication latency then we plan to dead-reckon the OptiTrack measurement using IMU, what do you think?

If you want to get the data from your OptiTrack system you can hook into the [extPositionHandler()](https://github.com/bitcraze/crazyflie-firmware/blob/e6ca813ae493b78a66a1e9ffa5585cbc051f0181/src/modules/src/crtp_localization_service.c#L180) funciton. Please see #661 for more information.

> And what is the best practice to get the high-accuracy high-frequency full-state measurement?
Should we use cflib or crazyswarm for that?

Using the ext_pose api. Both cflib or crazyswarm should work, but Crazyswam has a better integration with Optitrack, so it is maybe easier.

> Is there possible to build a very optimized program (no assertion, printing, overhead, etc)?

Yes, in firmware. You usually don't have to worry about asserts, but you should probably avoid prints.

**Reply by xkhainguyen** [reply]
Thank you for your response. 

> Using the ext_pose API.

Are you suggesting that I should handle the measurement by myself, not with the default setting (Kalman Filter maybe)?

> Yes, in firmware. You usually don't have to worry about asserts, but you should probably avoid prints.

How can I do that specifically? Because I feel like there is no option like that in the Kbuild or Makefile.

**Reply by krichardsson** [reply]
> Are you suggesting that I should handle the measurement by myself, not with the default setting (Kalman Filter maybe)?

It depends on what you want to do. The exp_pose API is mainly used to send position information from a mocap system to the Crazyflie. The default behavior is that the information is used by the kalman estimator to estimate the state. There is no example for Optitrack, [but one for Qualisys](https://github.com/bitcraze/crazyflie-lib-python/blob/6370208d61cd959c1009971558b53adb43e3a24b/examples/qualisys/qualisys_hl_commander.py#L208-L219) that you can get inspiration from. If you want to use Crazyswarm, there is already support for Optitrack.

> How can I do that specifically? Because I feel like there is no option like that in the Kbuild or Makefile.

Check out the links above for information on how to write an app and add an out-of-tree controller . You can also look at [the app examples](https://github.com/bitcraze/crazyflie-firmware/tree/master/examples)

**Comment by xkhainguyen** [comment]
Hello,
I can develop an LQR controller as in app layer (without additional library/sub-directory). However, when I try to integrate my library, it is quite tricky with Kbuild and make files.
My library looks like this:
```
.
└── my_controller/
    ├── CMakeLists.txt
    ├── src/
    │   ├── c1.h
    │   ├── c1.c
    │   ├── c2.h
    │   ├── c2.c
    │   ├── controller.h
    │   └── CMakeLists.txt
    └── external/
        └── linear_algebra_lib/
            ├── CMakeLists.txt
            └── src/
                ├── la1.h
                ├── la1.c
                ├── la2.h
                ├── la2.c
                ├── linear_algebra.h
                └── CMakeLists.txt
```
1/ Am I supposed to include all the output .o files in Kbuild?
2/ Is there a better way to modify make files to be compatible with my CMakeLists.txt?
3/ How about other options, like build a static/dynamic library?

**Reply by knmcguire** [reply]
Hi! Since you already selected an answer for this thread, I would like you to start a new discussion for this.
